fplll/hnf.h

```diff
@@ -0,0 +1,91 @@
+/* Copyright (C) 2018 Arnaud Sipasseuth.
+
+   This file is part of fplll. fplll is free software: you
+   can redistribute it and/or modify it under the terms of the GNU Lesser
+   General Public License as published by the Free Software Foundation,
+   either version 2.1 of the License, or (at your option) any later version.
+
+   fplll is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public License
+   along with fplll. If not, see <http://www.gnu.org/licenses/>. */
+
+/**
+  * WARNING : all HNFs here are considered lower triangular. Not upper.
+  */
+
+#ifndef FPLLL_HNF_H
+#define FPLLL_HNF_H
+
+#include "nr/matrix.h"
+
+FPLLL_BEGIN_NAMESPACE
+
+/**
+   @brief Test the HNF form of B.
+
+   @param A
+   @return zero on success.
+*/
+
+int is_hnf_reduced(const ZZ_mat<mpz_t> &B);
+
+/**
+ * @brief Tests a couple matrix vector
+ *    returns 1 if B is not a HNF or -1 if v is not in L(B), else 0
+ *
+ * @param B
+ *    HNF basis of the lattice to be tested
+ * @param v
+ *    vector for the membership test
+ */
+
+int in_hnf(const ZZ_mat<mpz_t> &B, const vector<Z_NR<mpz_t>> &v);
```

Comment:
I find `in_hnf` misleading, why not ask `in_lattice`?

Reply:
Because this only works if B is given is a HNF, but my naming is indeed confusing. I didn't want to include the HNF computation part. I will rename this to in_lattice_given_HNF. Makes things clearer.

Reply:
How expensive is it to test `is_hnf_reduced`?